logs/client.go

```diff
@@ -0,0 +1,199 @@
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	logsUrl, err := url.Parse(logsEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.opts.group != "" {
+		params.Add("group", c.opts.group)
+	}
+	if c.opts.minTime != "" {
+		params.Add("startTime", c.opts.minTime)
+	}
+	if c.opts.maxTime != "" {
+		params.Add("endTime", c.opts.maxTime)
+	}
+
+	var filter string
+	if c.opts.system != "" {
+		filter = fmt.Sprintf(`host:"%s"`, c.opts.system)
+	}
+	if len(c.opts.args) != 0 {
+		if len(filter) == 0 {
+			filter = strings.Join(c.opts.args, " ")
+		} else {
+			filter = filter + " " + strings.Join(c.opts.args, " ")
+		}
+	}
+
+	if filter != "" {
+		params.Add("filter", filter)
+	}
+
+	logsUrl.RawQuery = params.Encode()
+	request, err := http.NewRequestWithContext(ctx, "GET", logsUrl.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	request.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.opts.Token))
+	request.Header.Add("Accept", "application/json")
+
+	return request, nil
+}
+
+func (c *Client) printResult(logs []Log) error {
+	if c.opts.json {
+		jsonFormat, err := json.Marshal(logs)
+		if err != nil {
+			return err
+		}
+
+		_, err = fmt.Fprintln(c.output, string(jsonFormat))
+		return err
+	}
+
+	for i := len(logs) - 1; i >= 0; i-- {
+		l := logs[i]
+		fmt.Fprintf(c.output, "%s %s %s %s\n", l.Time.Format("Jan 02 15:04:05"), l.Hostname, l.Program, l.Message)
+	}
+
+	return nil
+}
+
+func (c *Client) Run(ctx context.Context) error {
+	if c.opts.version {
+		fmt.Fprintln(c.output, version.Version)
+		return nil
+	}
+
+	var logsCount int
+	var allLogs []Log
+	var nextPage string
+
+	for logsCount < int(c.opts.count) {
+		request, err := c.prepareRequest(ctx, nextPage)
+		if err != nil {
+			return fmt.Errorf("error while preparing http request to SWO: %w", err)
+		}
+
+		response, err := c.httpClient.Do(request)
+		if err != nil {
+			return fmt.Errorf("error while sending http request to SWO: %w", err)
+		}
+		defer func() {
+			err := response.Body.Close()
+			if err != nil {
+				slog.Error("Could not close https body", "error", err)
+			}
+		}()
+
+		content, err := io.ReadAll(response.Body)
+		if err != nil {
+			return fmt.Errorf("error while reading http response body from SWO: %w", err)
+		}
+
+		if !(response.StatusCode >= 200 && response.StatusCode < 300) {
+			return fmt.Errorf("received %d status code, response body: %s", response.StatusCode, string(content))
+		}
+
+		if len(content) == 0 {
+			return nil
+		}
+
+		var logs LogsData
+		err = json.Unmarshal(content, &logs)
+		if err != nil {
+			return fmt.Errorf("error while unmarshaling http response body from SWO: %w", err)
+		}
+
+		lastIdx := len(logs.Logs)
+		if logsCount+len(logs.Logs) > int(c.opts.count) {
+			lastIdx = int(c.opts.count) - logsCount
+		}
+		allLogs = append(allLogs, logs.Logs[:lastIdx]...)
+		logsCount += len(logs.Logs)
+
+		if logs.NextPage == "" {
+			break
+		}
+
+		nextPage = logs.NextPage
+	}
+
+	return c.printResult(allLogs)
```

Comment:
We should also be printing results as they come, to avoid having to buffer them all internally.

The API seems to return pages in reverse chronological order though. Maybe this is needed as a temporary measure until we can correct that.

Reply:
Yeah, that seems like a better way. What should we do when `--json` flag is present? Should we do some printing and emulate one big json logs array or print several log arrays?

Reply:
Changed json printing to papertrail-cli style

Reply:
also logs are no longer buffered before printing in order to print them in reverse